### PR TITLE
Fix issue with VRG protection causing Noobaa failures

### PR DESCRIPTION
**Pull request by BenamarMk**
After reinvestigating this [BZ](https://bugzilla.redhat.com/show_bug.cgi?id=2158728), we found that repetitive protection of VRG within a second caused Noobaa to start failing. To address this issue, this PR introduces throttling of VRG protection. Specifically, the function now allows for one protection per VRG in no less than 5 minutes, preventing over-protection that could lead to Noobaa failures. 

We were able to reproduce the issue by instrumenting Ramen to repeatedly protect the same VRG twice a second. After a few iterations, we started encountering the following error:
```
"err": "failed to upload data of odrbucket-81fc7663892a:busybox-workloads-3/busybox-drpc/*v1alpha1.VolumeReplicationGroup/b, InternalError: We encountered an internal error. Please try again.\n\tstatus code: 500, request id: lhhybhqo-g76y89-193, host id: lhhybhqo-g76y89-193"}
```

### CODE REVIEW

**Review by ShyamsundarR — APPROVED** [review]

**Review by hatfieldbrian — COMMENTED** [review]
There are multiple reasons the VRG is updated in S3:
- To update `ProtectedVolumeReplicationsGroup` list with changes to VRG, e.g., `spec` or `annotations`
- To update `VRG.status.kubeObjectProtection.captureToRecoverFrom` identifier.  There are two backup slots numbered 0 and 1, this field points to the latest complete backup.  By delaying VRG upload it would result in increasing Kube objects recovery point, which is 5 minutes by default, but a lower value may be specified in `VRG.spec.kubeObjectProtection.captureInterval`.  But more importantly, the caller needs to know if the pointer is not uploaded, so it doesn't think the pointer has advanced and proceed to the next backup.  Otherwise, recovery could fail by recovering from the wrong backup which is either missing or partial.  This is a known issue #844 which this patch does not introduce but does expose.

The first reason is likely the one causing to the noobaa issues.  I suggest it be optimized first, by uploading VRG only if its `resourceVersion` is different from the last one uploaded.

I also wonder if this noobaa issue could be addressed in other ways, e.g.,
- Understand why the excessive VRG reconciles and possibly reduce them by waiting for events instead timed backoff requeues
- Understand why noobaa does not tolerate the workload.  I imagine it designed for much more than Ramen gives it.  Perhaps it could be configured to tolerate more.

**Comment by ShyamsundarR** [comment]
> There are multiple reasons the VRG is updated in S3:
> 
>     * To update `ProtectedVolumeReplicationsGroup` list with changes to VRG, e.g., `spec` or `annotations`
> 
>     * To update `VRG.status.kubeObjectProtection.captureToRecoverFrom` identifier.  There are two backup slots numbered 0 and 1, this field points to the latest complete backup.  By delaying VRG upload it would result in increasing Kube objects recovery point, which is 5 minutes by default, but a lower value may be specified in `VRG.spec.kubeObjectProtection.captureInterval`.  But more importantly, the caller needs to know if the pointer is not uploaded, so it doesn't think the pointer has advanced and proceed to the next backup.  Otherwise, recovery could fail by recovering from the wrong backup which is either missing or partial.  This is a known issue [Kube objects pointer advance failure results in failed recovery #844](https://github.com/RamenDR/ramen/issues/844) which this patch does not introduce but does expose.

For the former, we can leverage a cached generation of the VRG to see if we need to upload it again or not. For the latter we would be in the same trouble as lastSyncTime is updated quite frequently in this case (read below)

> 
> 
> The first reason is likely the one causing to the noobaa issues. I suggest it be optimized first, by uploading VRG only if its `resourceVersion` is different from the last one uploaded.
> 
> I also wonder if this noobaa issue could be addressed in other ways, e.g.,
> 
>     * Understand why the excessive VRG reconciles and possibly reduce them by waiting for events instead timed backoff requeues

The excessive reconciles was due to 3 VRGs each protecting ~20 PVCs and their VR lastSyncTime updates, so basically the s3 load was ~750 PUTs in 30 minutes (not something that should break things, but it does)

> 
>     * Understand why noobaa does not tolerate the workload.  I imagine it designed for much more than Ramen gives it.  Perhaps it could be configured to tolerate more.

The manner to enable it to tolerate more was to scale the nooba service, which itself has issues of resource consumption etc.

**Comment by hatfieldbrian** [comment]
> For the former, we can leverage a cached generation of the VRG to see if we need to upload it again or not. For the latter we would be in the same trouble if we depend on resource version (maybe a little less than what it is now) as lastSyncTime is updated quite frequently in this case (read below)
> 

The reason I suggested `resourceVersion` instead of `generation` was to include `metadata` as I understand `generation` is only updated when `spec` changes and the requirement is to replicate `spec` and `metadata` for case one and `status` for case two.  Currently it does not upload based on reason so that would need to be added.

> basically the s3 load was ~750 PUTs in 30 minutes (not something that should break things, but it does)

Any indication of which resource was a bottleneck?

> The manner to enable it to tolerate more was to scale the nooba service, which itself has issues of resource consumption etc.

I wonder if an instance could be scaled up instead of out.  Maybe more memory, CPU, or storage capacity?

**Comment by BenamarMk** [comment]
> To update ProtectedVolumeReplicationsGroup list with changes to VRG, e.g., spec or annotations
> To update VRG.status.kubeObjectProtection.captureToRecoverFrom identifier. There are two backup slots numbered
I thought the KubeObjects protection is set at 5 minutes (defaulted), that's why I aimed for the 5 minutes.  What good does it do if the VRG is uploaded every second, but the KubeObjects are uploaded every 5 minutes?  They should be aligned, don't they? With this change, the recovery point for the VRG is always going to be 5min max. I don't see an issue here unless the KubeObjects backup schedule is less than 5 minutes. And in production, I don't think that's a good idea to have a KubeObjects backup that frequently.  

@ShyamsundarR suggested using a cache that can be populated every time Ramen restarts, but this doesn't completely solve the problem. In a heavily-loaded system with many PVCs per workload, changes are likely to occur frequently, unless we only check the diff against the Metadata section and the Spec section but not the status section. In that case, using the cache would defeat its purpose with added complexity.

> I wonder if an instance could be scaled up instead of out. Maybe more memory, CPU, or storage capacity?
I don't believe that scaling up an instance would be a solution to the current issue we are facing. Currently, we are causing excessive load on Noobaa by spamming it with unneeded frequent uploads, which is causing it to fail. While scaling up an instance might solve this temporarily, it would not address the root cause of the problem. If we need to upload data every second due to frequent changes to the object, then that might be a fair ask. For now, it's crucial to avoid overloading Noobaa with unnecessary uploads.

**Comment by hatfieldbrian** [comment]
@BenamarMk I agree, unnecessary uploads should be removed.  One way to do so is remove unnecessary reconciles.  For example, https://github.com/RamenDR/ramen/pull/883 proposes not reconciling VRG for VRG status and finalizer updates.  Another way to reduce uploads is to make them conditional.  1 PUT/5 minutes throttling is one way.  Detecting when something changes is another.

I think noobaa's capacity needs to be understood.  What is its write bandwidth?  What is its read bandwidth?  When does it break?  What other workloads could break it?  E.g., failover of several VRGs and PVCs due to disaster does at least one GET per PVC in a short amount of time.  For comparison, the real API server, limited as it is, does not seem to be breaking with a similar quantity of VRG PUTs and GETs as the VRG PUTs to noobaa.

**Comment by hatfieldbrian** [comment]
Ramen's S3 store requirements should also be defined, e.g., PUT and GET rate per VRG and per PVC.

File: controllers/vrg_vrgobject.go [hunk 1]
```diff
@@ -4,16 +4,40 @@
...
+	"time"
+
 	ramen "github.com/ramendr/ramen/api/v1alpha1"
 	"github.com/ramendr/ramen/controllers/util"
 	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	ctrl "sigs.k8s.io/controller-runtime"
 )
 
+var VRGLastUploadTime = map[string]metav1.Time{}
```
[the remark this answers is not in the document]
Comment (BenamarMk) [reply controllers/vrg_vrgobject.go:16]: sure

File: controllers/vrg_vrgobject.go [hunk 2]
```diff
@@ -4,16 +4,40 @@
...
+
+//nolint:gomnd
 func (v *VRGInstance) vrgObjectProtect(result *ctrl.Result, s3StoreAccessors []s3StoreAccessor) {
 	vrg := v.instance
 	eventReporter := v.reconciler.eventRecorder
 	log := v.log
+	key := vrg.Namespace + "." + vrg.Name
```
[the remark this answers is not in the document]
Comment (BenamarMk) [reply controllers/vrg_vrgobject.go:23]: A `dot` or a `slash` doesn't mean anything in this case. This is just the key to the last upload time map. Even if there is a resource with a dot in the name, it won't cause any harm.

File: controllers/vrg_vrgobject.go [hunk 3]
```diff
@@ -4,16 +4,40 @@
...
+
+	if lastUploadTime, ok := VRGLastUploadTime[key]; ok {
+		kubeObjectCaptureInterval := ramen.KubeObjectProtectionCaptureIntervalDefault
+		if vrg.Spec.KubeObjectProtection != nil {
+			kubeObjectCaptureInterval = kubeObjectsCaptureInterval(vrg.Spec.KubeObjectProtection)
+		}
+
+		// The maxVRGProtectionInterval is half the interval of the kubeObjectsCaptureInterval
+		maxVRGProtectionInterval := kubeObjectCaptureInterval / (2 * time.Nanosecond)
```
[the remark this answers is not in the document]
Comment (BenamarMk) [reply controllers/vrg_vrgobject.go:32]: This is just to make it readable given that the `KubeObjectProtectionCaptureIntervalDefault` is hardcoded as nanoseconds. I can remove it to make it less confusing given that the intention was make it not confusing.

**Comment by BenamarMk** [comment]
@hatfieldbrian as far as this one is concerned:
```
A VRG's VRGLastUploadTime entry should be deleted on VRG delete to avoid a memory leak
```
I am not worried about it.  How many VRGs are we creating before Ramen Operator restarts? It is a tolerable leak.

**Review by ShyamsundarR — APPROVED** [review]

**Pull request merged**